### ISSUE TITLE
Firebase Emulator simply does not initialize .env || .env.local || .env.default

### ISSUE THREAD

**Issue by polisen** [firebase/firebase-functions#1094]
### Related issues

https://github.com/firebase/firebase-functions/issues/1084

### [REQUIRED] Version info

**node:**v14.17.6

**firebase-functions:**3.20.1
**firebase-tools:**10.6.0

**firebase-admin:**10.1.0
### [REQUIRED] Test case

- Create local firebase project
- insert and populate .env, .env.local, .env.default (for good measure)
- start emulator suite
- log process.env

### [REQUIRED] Expected behavior

For process.env to have been populated with local .env values.

### [REQUIRED] Actual behavior
.env values are nowhere to be found.

### Were you able to successfully deploy your functions?
Technically, yes.

**Comment by taeold** [comment]
One issue I've seen with other customer is that the `.env` files were located outside the functions directory, e.g.

```
repo/
  firebase.json
  functions/
    .env
    .env.local
```

Can you share your directory setup to confirm?

**Comment by polisen** [comment]
I can confirm that the folder is located within functions.
The .env format is also correct.
I have rimraffed and re-installed everything, reconfigured emulators, removed .runtimeconfig.json but no dice.

**Comment by Verbunix** [comment]
Thanks, everyone for your research. I faced the same issue when locally "firebase-functions" did not get the environment variables.
Maybe this will help somebody while we are waiting to fix this issue:
1. Install [dotenv](https://www.npmjs.com/package/dotenv). `npm install dotenv --save`
2. [Use dotenv](https://github.com/motdotla/dotenv#how-do-i-use-dotenv-with-import) package in your function(for instance in index.ts) to load environment variables:
```
import { api } from './http';
import dotenv from 'dotenv';

dotenv.config();

export {
  api,
}
```
Now you can access your `.env` using `process.env.AWESOME_VAR_NAME`

**Comment by hellsingblack** [comment]
not working with firebase-tools 11, process.env doesn't set at all event if the .env is under functions/.env

**Comment by taeold** [comment]
@Verbunix Thanks for the hint. It sounds like the Emulator is silently ignoring badly configured environment variables. That's a pain - I'll follow up here if I can improve that.

**Comment by calclavia** [comment]
Seems related to this issue: https://stackoverflow.com/questions/71254427/firebase-functions-dotenv-variable-undefined

**Comment by Sutil** [comment]
The .env is not loaded when using the --import option.
`firebase emulators:start --export-on-exit --import <my-folder>`.

When using only `firebase emulators:start`, it is working fine.

**Comment by rishabh-rastogi215** [comment]
> Seems related to this issue: https://stackoverflow.com/questions/71254427/firebase-functions-dotenv-variable-undefined

@Sutil I was able to fix this by setting the same node version in package.json as you have in terminal. Also ensure that your package.json is using the same firebase-tools version. Once I was running the latest, it all worked

**Comment by Zachcodes** [comment]
Is there any news on this? I'm pretty sure this is still an issue.

**Comment by joshorvis** [comment]
I'm having the same problem - I'm using `firebase emulators:start --import=myImportFolder --export-on-exit` .

When I run this from my main project folder (C:/myProject) I am able to import and export my data successfully but none of my .env.local variables are available.  

If I `cd functions/` (C:/myProject/functions) and execute the same command there, I get all my variables BUT I'm not able to export successfully.    When the emulator shuts down, it deletes the import folder and creates TWO firebase-export-{alphanumericString} folders, both appear to be a copy of the exported data.  

Also worth noting that if I try to run `firebase emulators:export {importFolder}` while running in the functions folder, I get this error in the emulator logs (does not happen when running in project root):  
`!  emulators: Export failed: EPERM: operation not permitted, rename 'firebase-export-1667420687188p1aAWH' -> 'C:\myProject\functions\myImportFolder'`

**Comment by rscotten** [comment]
Confirmed this is still an issue.  

I'm running the firebase-tools 11.24.0.  

The Functions Emulator would not load the `.env` files into `process.env` before the function bootstraps even when I see this logged to the console:

```sh
functions: Loaded environment variables from .env, .env.<projectId>, .env.local. 
```

In some places it worked if I added this to my `./index.js` root/entry file:

```ts
import dotenv from 'dotenv';

dotenv.config();
```

But this doesn't work:

```ts
export const api = functions
  .region(process.env.region)
  .https.onRequest(app);
```

Which seems to defeat the purpose of the autoloading mechanism that's documented here:

https://firebase.google.com/docs/functions/config-env#emulator_support

and here:

https://firebase.google.com/docs/emulator-suite/connect_functions#configure_a_local_testing_environment

**Comment by nerder** [comment]
I've encountered the same issue when ` --import` is used. Only `.env` is loaded, `.env.local` is ignored. Any workaround?

**Comment by Padreco-lbr** [comment]
@polisen Can you leave a comment here so the bot will reopen the issue?
I can confirm it is still happening with firebase-functions@4.2.1 and firebase-tools@11.25.2

**Comment by jhonatanoliveira** [comment]
I had this issue of variables not being loaded, but my case was related to a misunderstanding with the doc.
Both the `.env` and `.env.dev` files are loaded, but the behavior of variable replacement is not what I expected.
TL;DR: To fix the issue, I defined all variables in `.env` and redefined them accordingly in `.env.dev`.

I understood that different env files are replaced depending on the project alias. For example, `.env.dev` would replace `.env` if I am in `firebase use dev`.
But this is not the case: a variable in `.env.dev` only replaces a variable from `.env` if it exists in `.env` in the first place.

Previously, my understanding was of a union behavior: variables from `.env` are loaded, then variables from `.env.dev` are loaded, and they replace existent variables from `.env`.
Now, my understanding is as follows: variables from `.env` are loaded and only this set of loaded variables can be redefined in `.env.dev`, no new variables can be added.

**Comment by juane1000** [comment]
> I had this issue of variables not being loaded, but my case was related to a misunderstanding with the doc. Both the `.env` and `.env.dev` files are loaded, but the behavior of variable replacement is not what I expected. TL;DR: To fix the issue, I defined all variables in `.env` and redefined them accordingly in `.env.dev`.
> 
> I understood that different env files are replaced depending on the project alias. For example, `.env.dev` would replace `.env` if I am in `firebase use dev`. But this is not the case: a variable in `.env.dev` only replaces a variable from `.env` if it exists in `.env` in the first place.
> 
> Previously, my understanding was of a union behavior: variables from `.env` are loaded, then variables from `.env.dev` are loaded, and they replace existent variables from `.env`. Now, my understanding is as follows: variables from `.env` are loaded and only this set of loaded variables can be redefined in `.env.dev`, no new variables can be added.

If this is true, can you write the docs instead? I'll test this out later tonight

**Comment by vdegenne** [comment]
@polisen Can you please re-open this issue? The bot closed it for inactivity but since then more people complained, including me.

I'm having the same problem. I opened a [SOF issue there](https://stackoverflow.com/questions/76493286/cant-use-process-env-values-in-cloud-functions-arguments) and then I found this GH issue so I thought I would share.
The problem seems to be that environment variables are not being loaded during the deployment process.  We should have a way to tell the deployment process to load the dotenv files.

(BTW @nerder and @jhonatanoliveira you guys are talking about a very different problem...)

**Comment by Diesmo** [comment]
My testing so far has shown that all variabel I define inside my .env files are available inside my function bodys, but only inside them.

So if I have an condition in my index.ts, where the function exports are defined, then process.env will only log the google firebase environment variables. But if I log the same process.env again inside on of my function and trigger it, then the correct and expected .env content will be printed.

This is somewhat understandable, as firebase does only include the .env file(s) in the package that gets uploaded and deployed. So the actual building step uses the host system variables that have startet the build/deploy script.
Be it a pipeline or the local shell on your computer.

But now the strange, and to me unexplainable, part starts.

If I run for example this command:
`export NODE_ENV=production && export TEST=test && export STAGE=stage && firebase emulators:start --only functions`
and put this logging near the top of my index.ts file:
```ts
console.log('NODE_ENV', process.env.NODE_ENV )
console.log('TEST', process.env.TEST )
console.log('STAGE', process.env.STAGE )
```

this is the output:
```
NODE_ENV production
TEST undefined
STAGE undefined
```

and if I only start it without any exports, like so:
`firebase emulators:start --only functions`

then the output changes to this:
```
NODE_ENV undefined
TEST undefined
STAGE undefined
```

I have no idea why only the NODE_ENV survives and everything else gets lost. Given, there are still a lot of firebase related variabel if I print my whole `process.env` but I have not idea where to look further for some explanation why my host system environment variabel are mostly ignored, apart from NODE_ENV.

I also tested to start it with a random NODE_ENV value, to make sure its not from some build process and prue coincidence:
`export NODE_ENV=foobar && export TEST=test && export STAGE=stage && firebase emulators:start --only functions`

That resulted in this output:
```
NODE_ENV foobar
TEST undefined
STAGE undefined
```

If anyone has any clues on what is wrong here or how this can get fixed I would be very grateful.

**Comment by ekalosak** [comment]
I can validate this. The simplest reproducible case is:

```py
# main.py
import os
if not os.getenv("WORKS"):
    raise ValueError("This won't get raised")
if not os.getenv("FAILS"):
    raise ValueError("This does get raised")
```

```
# .env
WORKS=1
```

```
# .env.local
FAILS=1
```

---
@Diesmo As a clue, the secret injection is also broken. I suuuuspect for the same reason: the environment variable injection system is bricked. I've been hunting around for the relevant source code, but only in spare time. Please update here if you find something.

**Comment by OskarGroth** [comment]
Running into the same issue here.
